WcaOnRails/spec/controllers/results_submission_controller_spec.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ResultsSubmissionController, type: :request do
+  let(:delegate) { FactoryBot.create :delegate }
+  let(:comp) { FactoryBot.create(:competition, delegates: [delegate]) }
+
+  context "not logged in" do
+    it "redirects to sign in" do
+      get submit_results_edit_path(comp.id)
+      expect(response).to redirect_to(new_user_session_path)
+    end
+  end
+
+  context "logged in as a regular user" do
+    sign_in { FactoryBot.create(:user) }
+
+    it "redirects to home page" do
+      get submit_results_edit_path(comp.id)
+      expect(response).to redirect_to(root_url)
+    end
+  end
+
+  context "logged in as a regular delegate" do
+    sign_in { FactoryBot.create(:delegate) }
+
+    it "redirects to home page" do
+      get submit_results_edit_path(comp.id)
+      expect(response).to redirect_to(root_url)
+    end
+  end
+
+  context "logged in as THE delegate" do
+    let(:user) { comp.delegates.first }
+    let(:submission_message) { "Hello, here are the results" }
+    let(:file_contents) { '{ "results": "good" }' }
+    let(:file) do
+      temp_file = Tempfile.new("sometmpfilename.tmp")
+      temp_file.write(file_contents)
+      temp_file.rewind
+      return Rack::Test::UploadedFile.new(temp_file.path, "application/json")
+    end
+
+    before :each do
+      sign_in user
+    end
+
+    describe "Seeing results submission page" do
+      it "returns http success" do
+        get submit_results_edit_path(comp.id)
+        expect(response).to have_http_status(:success)
+      end
+    end
+
+    # https://relishapp.com/rspec/rspec-mocks/v/3-2/docs/setting-constraints/matching-arguments#using-a-custom-matcher
+    RSpec::Matchers.define :file_containing do |contents|
+      match { |file| file.read == contents }
+    end
+
+    describe "Posting results" do
+      it "sends the 'results submitted' email immediately" do
+        expect(CompetitionsMailer)
+          .to receive(:results_submitted)
+          .with(comp, submission_message, file.original_filename, file_containing(file_contents))
+          .and_call_original
+        post submit_results_path(comp.id), params: { competition_id: comp.id, message: submission_message, results: file }
+
+        assert_enqueued_jobs 0
+      end
+
+      it "redirects to competition page" do
+        post submit_results_path(comp.id), params: { competition_id: comp.id, message: submission_message, results: file }
+
+        expect(flash[:success]).not_to be_empty
+        expect(response).to redirect_to(competition_path(comp))
+      end
+    end
+
+    describe "Posting results with missing message" do
+      it "flashes an error and doesn't send an email" do
+        post submit_results_path(comp.id), params: { competition_id: comp.id, results: file }
+
+        expect(flash.now[:danger]).not_to be_empty
+        assert_enqueued_jobs 0
```

Comment:
This is a bit tricky. This test is trying to verify that no email was sent, so it asserts that no jobs were enqueued.  However, the tests above that check that an email was in fact sent also check that no jobs were enqueued (because we expect to send emails immediately, not in a delayed job)! Unfortunately, I think the only way for us to gain confidence that emails were not sent would be to use a mock to assert that `CompetitionsMailer.results_submitted` was *not* called during this test.

Reply:
Maybe something like this?

```ruby
expect {
  post submit_results_path(comp.id), params: { competition_id: comp.id, results: file }
}.to_not change { ActionMailer::Base.deliveries.count }

Reply:
If @jonatanklosko's example works, I think that's by far a superior solution to my suggestion =)

Reply:
I see what you mean. Elegant solution, it works :)
I made the change in the test below as well.